File: code/php/rest/server/src/Services/TaskService.php
<?php
declare(strict_types=1);

namespace TaskAPI\Services;

use TaskAPI\Models\Task;
use TaskAPI\Exceptions\NotFoundException;
use TaskAPI\Exceptions\ValidationException;

class TaskService
{
    private array $tasks = [];
    
    public function __construct()
    {
        $this->initializeSampleData();
    }
    
    public function listTasks(
        int $pageSize = 20,
        ?string $pageToken = null,
        ?string $status = null,
        ?string $assignedTo = null,
        ?string $tags = null,
        ?string $sortOrder = null
    ): array {
        // Filter tasks
        $filteredTasks = array_filter($this->tasks, function(Task $task) use ($status, $assignedTo, $tags) {
            return $task->matchesFilters($status, $assignedTo, $tags);
        });
        
        // Sort tasks
        $sortedTasks = $this->sortTasks(array_values($filteredTasks), $sortOrder);
        
        // Paginate
        $pageSize = min(max($pageSize, 1), 100);
        $startIndex = $pageToken ? (int)$pageToken : 0;
        $endIndex = $startIndex + $pageSize;
        
        $tasksPage = array_slice($sortedTasks, $startIndex, $pageSize);
        $nextPageToken = $endIndex < count($sortedTasks) ? (string)$endIndex : null;
        
        return [
            'tasks' => array_map(fn($task) => $task->jsonSerialize(), $tasksPage),
            'next_page_token' => $nextPageToken,
            'total_count' => count($sortedTasks)
        ];
    }
    
    public function getTask(string $id): Task
    {
        if (!isset($this->tasks[$id])) {
            throw new NotFoundException("Task with ID {$id} not found");
        }
        
        return $this->tasks[$id];
    }
    
    public function createTask(array $data): Task
    {
        $task = Task::fromArray($data);
        
        if (!$task->isValid()) {
            $errors = $task->getValidationErrors();
            throw new ValidationException(implode(', ', $errors));
        }
        
        $this->tasks[$task->getId()] = $task;
        
        return $task;
    }
    
    public function updateTask(string $id, array $data): Task
    {
        if (!isset($this->tasks[$id])) {
            throw new NotFoundException("Task with ID {$id} not found");
        }
        
        $task = $this->tasks[$id];
        $task->update($data);
        
        if (!$task->isValid()) {
            $errors = $task->getValidationErrors();
            throw new ValidationException(implode(', ', $errors));
        }
        
        return $task;
    }
    
    public function updateTaskStatus(string $id, string $status): Task
    {
        return $this->updateTask($id, ['status' => $status]);
    }
    
    public function deleteTask(string $id): void
    {
        if (!isset($this->tasks[$id])) {
            throw new NotFoundException("Task with ID {$id} not found");
        }
        
        unset($this->tasks[$id]);
    }
    
    private function initializeSampleData(): void
    {
        $sampleTasks = [
            [
                'title' => 'Implement PHP REST API',
                'description' => 'Create REST API using Slim Framework',
                'priority' => Task::PRIORITY_HIGH,
                'tags' => ['php', 'rest', 'api']
            ],
            [
                'title' => 'Add gRPC support',
                'description' => 'Implement gRPC server with PHP',
                'priority' => Task::PRIORITY_MEDIUM,
                'tags' => ['php', 'grpc', 'protobuf']
            ],
            [
                'title' => 'Write unit tests',
                'description' => 'Add PHPUnit tests for API endpoints',
                'priority' => Task::PRIORITY_HIGH,
                'tags' => ['php', 'testing', 'phpunit']
            ]
        ];
        
        foreach ($sampleTasks as $taskData) {
            $this->createTask($taskData);
        }
    }
    
    private function sortTasks(array $tasks, ?string $sortOrder): array
    {
        switch ($sortOrder) {
            case 'created_at_asc':
                usort($tasks, fn($a, $b) => strcmp($a->getCreatedAt(), $b->getCreatedAt()));
                break;
            case 'created_at_desc':
                usort($tasks, fn($a, $b) => strcmp($b->getCreatedAt(), $a->getCreatedAt()));
                break;
            case 'due_date_asc':
                usort($tasks, fn($a, $b) => strcmp($a->getDueDate() ?? '9999-12-31', $b->getDueDate() ?? '9999-12-31'));
                break;
            case 'due_date_desc':
                usort($tasks, fn($a, $b) => strcmp($b->getDueDate() ?? '0000-01-01', $a->getDueDate() ?? '0000-01-01'));
                break;
            case 'priority_asc':
                usort($tasks, fn($a, $b) => $a->getPriorityValue() <=> $b->getPriorityValue());
                break;
            case 'priority_desc':
                usort($tasks, fn($a, $b) => $b->getPriorityValue() <=> $a->getPriorityValue());
                break;
        }
        
        return $tasks;
    }
}

File: code/ruby/rest/server/services/task_service.rb
require_relative '../models/task'

module Services
  class TaskService
    def initialize
      @tasks = {}
      @mutex = Mutex.new
      initialize_sample_data
    end
    
    def list_tasks(page_size: 20, page_token: nil, status: nil, assigned_to: nil, tags: nil, sort_order: nil)
      @mutex.synchronize do
        # Filter tasks
        filtered_tasks = @tasks.values.select do |task|
          task.matches_filters?(status: status, assigned_to: assigned_to, tags: tags)
        end
        
        # Sort tasks
        sorted_tasks = sort_tasks(filtered_tasks, sort_order)
        
        # Paginate
        page_size = [[page_size.to_i, 1].max, 100].min
        start_index = page_token ? page_token.to_i : 0
        end_index = start_index + page_size
        
        tasks_page = sorted_tasks[start_index...end_index] || []
        
        next_page_token = end_index < sorted_tasks.length ? end_index.to_s : nil
        
        {
          tasks: tasks_page.map(&:to_h),
          next_page_token: next_page_token,
          total_count: sorted_tasks.length
        }
      end
    end
    
    def get_task(id)
      @mutex.synchronize do
        task = @tasks[id]
        raise NotFoundError, "Task with ID #{id} not found" unless task
        task.to_h
      end
    end
    
    def create_task(attrs)
      task = Models::Task.new(attrs)
      
      unless task.valid?
        raise ValidationError, task.validation_errors.join(', ')
      end
      
      @mutex.synchronize do
        @tasks[task.id] = task
      end
      
      task.to_h
    end
    
    def update_task(id, attrs)
      @mutex.synchronize do
        task = @tasks[id]
        raise NotFoundError, "Task with ID #{id} not found" unless task
        
        task.update!(attrs)
        
        unless task.valid?
          raise ValidationError, task.validation_errors.join(', ')
        end
        
        task.to_h
      end
    end
    
    def update_task_status(id, status)
      update_task(id, { status: status })
    end
    
    def delete_task(id)
      @mutex.synchronize do
        task = @tasks.delete(id)
        raise NotFoundError, "Task with ID #{id} not found" unless task
        nil
      end
    end
    
    private
    
    def initialize_sample_data
      sample_tasks = [
        {
          title: 'Implement Ruby REST API',
          description: 'Create REST API using Sinatra framework',
          priority: 'high',
          tags: ['ruby', 'rest', 'api']
        },
        {
          title: 'Add gRPC support',
          description: 'Implement gRPC server with grpc gem',
          priority: 'medium',
          tags: ['ruby', 'grpc', 'protobuf']
        },
        {
          title: 'Write tests',
          description: 'Add RSpec tests for API endpoints',
          priority: 'high',
          tags: ['ruby', 'testing', 'rspec']
        }
      ]
      
      sample_tasks.each { |attrs| create_task(attrs) }
    end
    
    def sort_tasks(tasks, sort_order)
      case sort_order
      when 'created_at_asc'
        tasks.sort_by(&:created_at)
      when 'created_at_desc'
        tasks.sort_by(&:created_at).reverse
      when 'due_date_asc'
        tasks.sort_by { |t| t.due_date || Time.new(9999) }
      when 'due_date_desc'
        tasks.sort_by { |t| t.due_date || Time.new(0) }.reverse
      when 'priority_asc'
        tasks.sort_by(&:priority_value)
      when 'priority_desc'
        tasks.sort_by(&:priority_value).reverse
      else
        tasks
      end
    end
  end
  
  class ValidationError < StandardError; end
  class NotFoundError < StandardError; end
end